packages/map-toolkit/src/viewport/use-viewport-state.ts

```diff
@@ -0,0 +1,221 @@
+ * @returns A snapshot function for useSyncExternalStore
+ */
+function getOrCreateSnapshot(instanceId: UniqueId): () => MapViewportPayload {
+  const fallback = { ...defaultSnapshot, id: instanceId };
+  const snapshot =
+    snapshotCache.get(instanceId) ??
+    (() => viewportStore.get(instanceId) ?? fallback);
+
+  snapshotCache.set(instanceId, snapshot);
+
+  return snapshot;
+}
+
+/**
+ * Hook to subscribe to viewport state changes for a specific view.
+ *
+ * By default, subscribes to MapEvents.viewport events from the event bus and
+ * automatically cleans up when all subscribers unmount. For advanced use cases,
+ * custom subscribe/getSnapshot functions can be provided.
+ *
+ * A thin wrapper around [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore).
+ *
+ * @param instanceId - Unique identifier for the viewport to track
+ * @param subscribe - Optional custom subscription function
+ * @param getSnapshot - Optional custom snapshot getter
+ * @param getServerSnapshot - Optional server-side snapshot getter
+ * @returns Current viewport state including bounds, latitude, longitude, zoom
+ *
+ * @example
+ * ```tsx
+ * function MapInfo({ instanceId }) {
+ *   const { bounds, latitude, longitude, zoom } = useViewportState({
+ *     instanceId
+ *   });
+ *
+ *   return (
+ *     <div>
+ *       Lat: {latitude?.toFixed(2)}, Lon: {longitude?.toFixed(2)}, Zoom: {zoom}
+ *     </div>
+ *   );
+ * }
+ * ```
+ *
+ * @example
+ * ```tsx
+ * // With custom subscribe/getSnapshot for advanced use cases
+ * function CustomMapInfo() {
+ *   const customSubscribe = (onStoreChange) => {
+ *     // Custom subscription logic
+ *     return () => { // cleanup }
+ *   }
+ *
+ *   const customGetSnapshot = () => {
+ *     // Custom snapshot logic
+ *     return { latitude: 0, longitude: 0, zoom: 1 };
+ *   };
+ *
+ *   const viewState = useViewportState({
+ *     instanceId: 'some-uuid',
+ *     subscribe: customSubscribe,
+ *     getSnapshot: customGetSnapshot,
+ *   });
+ *
+ *   return <div>Custom viewport state</div>;
+ * }
+ * ```
+ */
+export function useViewportState({
+  instanceId,
+  subscribe,
+  getSnapshot,
+  getServerSnapshot,
+}: UseViewportStateProps) {
+  return useSyncExternalStore<MapViewportPayload>(
+    subscribe ?? getOrCreateSubscription(instanceId),
+    getSnapshot ?? getOrCreateSnapshot(instanceId),
+    getServerSnapshot,
+  );
+}
+
+/**
+ * Manually clear viewport state for a specific instanceId.
+ * This is typically not needed as cleanup happens automatically when all subscribers unmount.
+ * Use this only in advanced scenarios where manual cleanup is required.
+ *
+ * @param instanceId - The unique identifier for the viewport to clear
+ *
+ * @example
+ * ```tsx
+ * // Manual cleanup (rarely needed)
+ * clearViewportState('my-map-instance');
+ * ```
+ */
+export function clearViewportState(instanceId: UniqueId): void {
+  viewportStore.delete(instanceId);
+  subscriberCounts.delete(instanceId);
```

Comment:
Also update this to reflect the new logic: 

```suggestion
  // Unsubscribe from bus if listening
  const unsub = busUnsubscribers.get(instanceId);
  if (unsub) {
    unsub();
    busUnsubscribers.delete(instanceId);
  }

  // Clear all state
  viewportStore.delete(instanceId);
  componentSubscribers.delete(instanceId);
```